### PR TITLE
add logstash logback plugin

### PR DESCRIPTION
**Pull request by wuxingye**
Please answer these questions before submitting pull request

- Why submit this pull request?
- [ ] Bug fix
- [ ] New feature provided
- [ ] Improve performance

- Related issues

___
### Bug fix
- Bug description.

- How to fix?

___
### New feature or improvement
- Describe the details and related test reports.

the console:
image

the elasticsearch data from logstash:

image

Let me know if it doesn't meet the standard, i never commit code to such a great project, thank you!

### CODE REVIEW

**Comment by wuxingye** [comment]
the skywalking ui
image

**Comment by wuxingye** [comment]
the logback-spring.xml file

```xml
<appender name="stash" class="net.logstash.logback.appender.LogstashTcpSocketAppender">
    <destination>${destination}</destination>
    <connectionStrategy>
        <roundRobin>
            <connectionTTL>5 seconds</connectionTTL>
        </roundRobin>
    </connectionStrategy>
    <reconnectionDelay>1 second</reconnectionDelay>
    <encoder charset="UTF-8" class="net.logstash.logback.encoder.LogstashEncoder">
        <provider class="org.apache.skywalking.apm.toolkit.log.logback.v1.x.logstash.TraceIdJsonProvider">
        </provider>
        <customFields>{"app_id":"${app_id}"}</customFields>
    </encoder>
</appender>
```

**Comment by wuxingye** [comment]
the maven dependency

```xml
            <dependency>
                <groupId>org.apache.skywalking</groupId>
                <artifactId>apm-toolkit-logback-1.x</artifactId>
                <version>${skywalking.version}</version>
            </dependency>
```

**Comment by kezhenxu94** [comment]
@wuxingye if possible, I think you can upload your test project to a GitHub repository instead of posting code snippets here, which should be more friendly to reviewers and we can verify your changes more easily

**Comment by wuxingye** [comment]
I'll do the rest of the work tomorrow, thank you

**Comment by wuxingye** [comment]
# Add logstash logback plugin

## 1.user guide

- Dependency the toolkit, such as using maven or gradle

```xml
<dependency>
    <groupId>org.apache.skywalking</groupId>
    <artifactId>apm-toolkit-logback-1.x</artifactId>
    <version>${skywalking.version}</version>
</dependency>
```

- set `LogstashEncoder` of logback.xml

```xml
<encoder charset="UTF-8" class="net.logstash.logback.encoder.LogstashEncoder">
    <provider class="org.apache.skywalking.apm.toolkit.log.logback.v1.x.logstash.TraceIdJsonProvider">
    </provider>
    <customFields>{"app_id":"${app_id}"}</customFields>
</encoder>
```



## 2.why it works

- extension of logstash-logback-encoder provider, [official guide](https://github.com/logstash/logstash-logback-encoder#custom-json-provider)
- Custom LogstashEncoder JSON Provider like this:

```xml
<encoder class="net.logstash.logback.encoder.LogstashEncoder">
    ...
    <provider class="your.provider.YourJsonProvider">
        <!-- Any properties exposed by your provider can be set here -->
    </provider>
    ...
</encoder>
```

- TraceIdJsonProvider analyze

```java
public class TraceIdJsonProvider extends AbstractFieldJsonProvider<ILoggingEvent> implements FieldNamesAware<LogstashFieldNames> {

    public static final String TRACING_ID = "TID";

    @Override
    public void writeTo(JsonGenerator generator, ILoggingEvent event) throws IOException {
        Map<String, String> map = event.getLoggerContextVO().getPropertyMap();
        JsonWritingUtils.writeStringField(generator, getFieldName(), map.get(TRACING_ID));
    }

    @Override
    public void setFieldNames(LogstashFieldNames fieldNames) {
        setFieldName(TRACING_ID);
    }
}
```

1. the setFieldNames method will add a json field named `TID`
2. the writeTo method will set the `TID` value, for `propertyMap` is always not null, it at least contains `destination` of logstash, so the code will not throw `NPE`

- TcpSocketAppenderActivation analyze

```java
public class TcpSocketAppenderActivation extends ClassInstanceMethodsEnhancePluginDefine {

    public static final String INTERCEPT_CLASS = "org.apache.skywalking.apm.toolkit.activation.log.logback.v1.x.logstash.TcpSocketAppenderInterceptor";
    public static final String ENHANCE_CLASS = "net.logstash.logback.appender.LogstashTcpSocketAppender";
    public static final String ENHANCE_METHOD = "prepareForDeferredProcessing";

    /**
     * @return the target class, which needs active.
     */
    @Override
    protected ClassMatch enhanceClass() {
        return byName(ENHANCE_CLASS);
    }

    /**
     * @return null, no need to intercept constructor of enhance class.
     */
    @Override
    public ConstructorInterceptPoint[] getConstructorsInterceptPoints() {
        return null;
    }

    /**
     * @return the collection of {@link StaticMethodsInterceptPoint}, represent the intercepted methods and their
     * interceptors.
     */
    @Override
    public InstanceMethodsInterceptPoint[] getInstanceMethodsInterceptPoints() {
        return new InstanceMethodsInterceptPoint[] {
            new InstanceMethodsInterceptPoint() {
                @Override
                public ElementMatcher<MethodDescription> getMethodsMatcher() {
                    return named(ENHANCE_METHOD).and(takesArgumentWithType(0, "ch.qos.logback.classic.spi.ILoggingEvent"));
                }

                @Override
                public String getMethodsInterceptor() {
                    return INTERCEPT_CLASS;
                }

                @Override public boolean isOverrideArgs() {
                    return false;
                }
            }
        };
    }
}

```

1. net.logstash.logback.appender.`LogstashTcpSocketAppender` is logstash appender, contains a method:

```java
    @Override
    protected void prepareForDeferredProcessing(final ILoggingEvent event) {
        super.prepareForDeferredProcessing(event);
        if (includeCallerData) {
            event.getCallerData();
        }
    }
```

2. so we enhance it by the `beforeMethod` of `TcpSocketAppenderInterceptor` :

```java
    @Override public void beforeMethod(EnhancedInstance objInst, Method method, Object[] allArguments,
        Class<?>[] argumentsTypes, MethodInterceptResult result) throws Throwable {
        ILoggingEvent event = (ILoggingEvent)allArguments[0];
        if (event != null && event.getLoggerContextVO() != null && event.getLoggerContextVO().getPropertyMap() != null) {
            event.getLoggerContextVO().getPropertyMap().put("TID", ContextManager.getGlobalTraceId());
        }
    }
```

3. it put a key value of `(TID, ContextManager.getGlobalTraceId())` to the `ILoggingEvent`, so we can get it in the `writeTo` method of `TraceIdJsonProvider` 
4. the method `prepareForDeferredProcessing` of `LogstashTcpSocketAppender` works on the worker thread, but others works on the logstash thread, because `logstash-logback-encoder` if a async plugin, so we enhance it by this way

## 3.related pictures

- ConsoleAppender plugin

image

- logstash logback plugin

image

- skywalking ui

image

- before add `TID`

image

- after add `TID`

image

**Comment by wuxingye** [comment]
@wu-sheng @kezhenxu94

File: apm-sniffer/apm-toolkit-activation/apm-toolkit-logback-1.x-activation/pom.xml [hunk 1]
```diff
@@ -27,5 +28,15 @@
...
 
+    <properties>
+        <logback-classic.version>1.2.3</logback-classic.version>
+    </properties>
 
+    <dependencies>
+        <dependency>
+            <groupId>ch.qos.logback</groupId>
+            <artifactId>logback-classic</artifactId>
+            <version>${logback-classic.version}</version>
```
Comment (wu-sheng) [inline apm-sniffer/apm-toolkit-activation/apm-toolkit-logback-1.x-activation/pom.xml]: Dependency should be provided only.

**Comment by wu-sheng** [comment]
@wuxingye You need to move the part of posted document, which related to the end user, to the official SkyWalking. I think it should be placed in the document of loglack application toolkit.

**Comment by wu-sheng** [comment]
Resolves #3348 3352

File: apm-sniffer/apm-toolkit-activation/apm-toolkit-logback-1.x-activation/pom.xml (hunk 1 above)
Comment (wuxingye) [reply apm-sniffer/apm-toolkit-activation/apm-toolkit-logback-1.x-activation/pom.xml]: done

**Comment by wuxingye** [comment]
> @wuxingye You need to move the part of posted document, which related to the end user, to the official SkyWalking. I think it should be placed in the document of loglack application toolkit.

I had refine the doc, pls check, thx

File: docs/en/setup/service-agent/java-agent/Application-toolkit-logback-1.x.md [hunk 2]
```diff
@@ -55,3 +56,169 @@
...
+```
+
+3. it put a key value of `(TID, ContextManager.getGlobalTraceId())` to the `ILoggingEvent`, so we can get it in the `writeTo` method of `TraceIdJsonProvider` 
+4. the method `prepareForDeferredProcessing` of `LogstashTcpSocketAppender` works on the worker thread, but others works on the logstash thread, because `logstash-logback-encoder` if a async plugin, so we enhance it by this way
+
+## 3.related pictures
+
+- ConsoleAppender plugin
+
+![image-20190829093550854](/Users/wuxingye/Library/Application Support/typora-user-images/image-20190829093550854.png)
```
Comment (kezhenxu94) [inline docs/en/setup/service-agent/java-agent/Application-toolkit-logback-1.x.md]: how could you expect the users to see these pictures 😢

File: apm-sniffer/apm-toolkit-activation/apm-toolkit-logback-1.x-activation/src/main/java/org/apache/skywalking/apm/toolkit/activation/log/logback/v1/x/logstash/TcpSocketAppenderActivation.java [hunk 3]
```diff
@@ -0,0 +1,88 @@
...
+import static org.apache.skywalking.apm.agent.core.plugin.bytebuddy.ArgumentTypeNameMatch.takesArgumentWithType;
+import static org.apache.skywalking.apm.agent.core.plugin.match.NameMatch.byName;
+
+/**
+ * Active the toolkit class "org.apache.skywalking.apm.toolkit.log.logback.v1.x.LogbackPatternConverter". Should not
+ * dependency or import any class in "skywalking-toolkit-logback-1.x" module. Activation's classloader is diff from
+ * "org.apache.skywalking.apm.toolkit.log.logback.v1.x.LogbackPatternConverter", using direct will trigger classloader
+ * issue.
+ * <p>
+ * Created by wusheng on 2016/12/7.
```
Comment (wu-sheng) [inline apm-sniffer/apm-toolkit-activation/apm-toolkit-logback-1.x-activation/src/main/java/org/apache/skywalking/apm/toolkit/activation/log/logback/v1/x/logstash/TcpSocketAppenderActivation.java]: This class is not created by me, the comment seems copied from old one.

File: docs/en/setup/service-agent/java-agent/Application-toolkit-logback-1.x.md [hunk 4]
```diff
@@ -55,3 +56,25 @@
...
+</dependency>
+```
+
+* set `LogstashEncoder` of logback.xml
+
+```xml
+<encoder charset="UTF-8" class="net.logstash.logback.encoder.LogstashEncoder">
+    <provider class="org.apache.skywalking.apm.toolkit.log.logback.v1.x.logstash.TraceIdJsonProvider">
+    </provider>
+    <customFields>{"app_id":"${app_id}"}</customFields>
```
Comment (wu-sheng) [inline docs/en/setup/service-agent/java-agent/Application-toolkit-logback-1.x.md]: What is `app_id`? Is it related to SkyWalking?

File: apm-application-toolkit/apm-toolkit-logback-1.x/src/main/java/org/apache/skywalking/apm/toolkit/log/logback/v1/x/logstash/TraceIdJsonProvider.java [hunk 5]
```diff
@@ -0,0 +1,48 @@
...
+
+    @Override
+    public void writeTo(JsonGenerator generator, ILoggingEvent event) throws IOException {
+        Map<String, String> map = event.getLoggerContextVO().getPropertyMap();
+        JsonWritingUtils.writeStringField(generator, getFieldName(), map.get(TRACING_ID));
+    }
+
+    @Override
+    public void setFieldNames(LogstashFieldNames fieldNames) {
+        setFieldName(TRACING_ID);
```
Comment (wu-sheng) [inline apm-application-toolkit/apm-toolkit-logback-1.x/src/main/java/org/apache/skywalking/apm/toolkit/log/logback/v1/x/logstash/TraceIdJsonProvider.java]: Why is this `fieldNames` used? There is a change to config the key, to override `TID`?

Comment (wu-sheng) [reply apm-application-toolkit/apm-toolkit-logback-1.x/src/main/java/org/apache/skywalking/apm/toolkit/log/logback/v1/x/logstash/TraceIdJsonProvider.java]: Any explanation about this?

Comment (wu-sheng) [reply apm-application-toolkit/apm-toolkit-logback-1.x/src/main/java/org/apache/skywalking/apm/toolkit/log/logback/v1/x/logstash/TraceIdJsonProvider.java]: Make sense.

**Comment by wu-sheng** [comment]
@wuxingye  All core team members have no experience with this. So we decided to accept this due to no obvious harm. If someone uses this in the future, we will ping you at GitHub, hope you could help.

**Comment by wuxingye** [comment]
> @wuxingye All core team members have no experience with this. So we decided to accept this due to no obvious harm. If someone uses this in the future, we will ping you at GitHub, hope you could help.

@wu-sheng Ok, it's my pleasure

**Comment by wu-sheng** [comment]
@wuxingye Look like your comment break javadoc generation, could you recheck?

**Comment by wuxingye** [comment]
> @wuxingye Look like your comment break javadoc generation, could you recheck?

@wu-sheng  I find an error and a warning, if you fell there are still mistakes, pls help me, thx

**Comment by wu-sheng** [comment]
I just find out from the CI, nothing more.

**Comment by wu-sheng** [comment]
Click `Details` following CI. You will have Jenkins logs, it shows the commands and logs. In the Apache, we are required to have javadoc.

**Comment by wuxingye** [comment]
It's my first PR at GitHub, i gained a lot, and i will pay attention to CI in the later work

**Comment by wu-sheng** [comment]
Take your time. SkyWalking is a large ecosystem, even PMC member is not familiar all parts of the project. So, don't worry. With your further contributions, you will learn more.

And welcome to be 151th contributor of this repo.

**Comment by wuxingye** [comment]
Thank you.

**Comment by Jayanzhou** [comment]
@wuxingye 大佬

**Comment by HuangDayu** [comment]
Hello , @wuxingye  @wu-sheng   can you help me ? thank you .  https://github.com/apache/skywalking/issues/5746

**Comment by zhangxiaojiawow** [comment]
@wuxingye If I want to use another appender `ch.qos.logback.core.rolling.RollingFileAppender`,  how to add TraceId in JSON format log.
the following config doesn't take effect:
```xml
<encoder charset="UTF-8" class="net.logstash.logback.encoder.LogstashEncoder">
    <provider class="org.apache.skywalking.apm.toolkit.log.logback.v1.x.logstash.TraceIdJsonProvider">
    </provider>
    <customFields>{"app_id":"${app_id}"}</customFields>
</encoder>
```

**Pull request merged**